rmw_dds_common/test/test_graph_cache.cpp

```diff
@@ -0,0 +1,1132 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <cstring>
+#include <string>
+#include <tuple>
+#include <utility>
+#include <vector>
+
+#include "rmw/qos_profiles.h"
+#include "rmw/topic_endpoint_info.h"
+#include "rmw/topic_endpoint_info_array.h"
+
+#include "rmw_dds_common/graph_cache.hpp"
+
+using rmw_dds_common::GraphCache;
+
+struct NameAndNamespace
+{
+  std::string namespace_;
+  std::string name;
+};
+
+MATCHER_P2(IsNameAndNamespace, namespace_, name, "")
+{
+  return (arg.namespace_ == namespace_) && (arg.name == name);
+}
+
+void
+check_names_and_namespace(
+  const rcutils_string_array_t & names,
+  const rcutils_string_array_t & namespaces,
+  const std::vector<NameAndNamespace> & expected)
+{
+  ASSERT_EQ(names.size, namespaces.size);
+  ASSERT_EQ(expected.size(), names.size);
+  for (size_t i = 0; i < expected.size(); i++) {
+    ASSERT_THAT(
+      expected,
+      testing::Contains(IsNameAndNamespace(namespaces.data[i], names.data[i])));
+  }
+}
+
+struct NameAndTypes
+{
+  std::string name;
+  std::vector<std::string> types;
+};
+
+void check_names_and_types(
+  const rmw_names_and_types_t & names_and_types,
+  const std::vector<NameAndTypes> & expected)
+{
+  ASSERT_EQ(names_and_types.names.size, expected.size());
+
+  for (size_t i = 0; i < expected.size(); i++) {
+    const auto & item = expected[i];
+    EXPECT_EQ(item.name, names_and_types.names.data[i]);
+    const auto & expected_types = item.types;
+    const auto & types = names_and_types.types[i];
+    ASSERT_EQ(expected_types.size(), types.size);
+    for (size_t j = 0; j < expected_types.size(); j++) {
+      EXPECT_EQ(expected_types[j], types.data[j]);
+    }
+  }
+}
+
+std::string
+identity_demangle(const std::string & name)
+{
+  return name;
+}
+
+using DemangleFunctionT = GraphCache::DemangleFunctionT;
+
+void
+check_results(
```

Comment:
@ivanpauno nit: for all `check_*` functions, consider not using default parameters for empty sets or zero counts and provide them explicitly instead. IMO it'll make the tests much easier to read.

Reply:
I'd also be a bit more verbose on the naming e.g. `check_results` -> ` check_graph_cache_state`. Same for the rest.

Reply:
Still applies, but won't block on it.